text/3338-style-evolution.md

```diff
@@ -0,0 +1,99 @@
+
+# Motivation
+[motivation]: #motivation
+
+The current Rust style, as defined in the Rust Style Guide and as implemented
+by rustfmt, has some stability expectations associated with it. In particular,
+many projects implement continuous integration hooks that verify the style of
+Rust code (such as with `cargo fmt --check`), and changes to the style would
+break the CI of such projects, in addition to causing churn.
+
+This document proposes to evolve the current Rust style, without breaking
+backwards compatibility, by tying style evolution to Rust edition. Code in Rust
+2015, 2018, or 2021 will use the existing default style. Code in future
+editions (Rust 2024 and onwards) may use a new style edition.
+
+This RFC only defines the mechanism by which we evolve the Rust style; this RFC
+does *not* define any specific style changes. Future RFCs or style-guide PRs
+will define future style editions. This RFC does not propose or define any
+specific future style editions or other formatting changes.
+
+# Explanation
+[explanation]: #explanation
+
+`rustfmt`, and `cargo fmt`, will format code according to the default Rust
+style. The default Rust style varies by Rust edition. (Not every edition
+changes the Rust style, and thus some editions have identical default styles;
+Rust 2015, 2018, and 2021 all have the same default style.)
+
+Direct invocations of `rustfmt` obtain the edition used for parsing Rust code
+from the `edition` option in its configuration file (`rustfmt.toml` or
+`.rustfmt.toml`), or via the `--edition` command-line option; `cargo fmt`
+obtains the edition from the `edition` option in `Cargo.toml` and passes it to
+`rustfmt`. By default, `rustfmt` and `cargo fmt` will use the same edition for
+style as the Rust edition used for parsing.
+
+However, when transitioning between editions, projects may want to separately
+make and commit the changes for 1) transitioning to a new Rust edition and 2)
+transitioning to a new style edition. To allow for this, `rustfmt` also allows
+configuring the style edition directly, via a separate `style_edition`
+configuration option, or `--style-edition` command-line option. `style_edition`
+or `--style-edition`, if set, always overrides `edition` or `--edition` for the
+purposes of styling, though `edition` or `--edition` still determines the
+edition for the purposes of parsing Rust code.
+
+Note that rustfmt may not necessarily support all combinations of Rust edition
+and style edition; in particular, it may not support using a style edition that
+differs by more than one step from the Rust edition.
+
+The current version of the style guide will describe the latest Rust edition.
+Each distinct past style will have a corresponding archived version of the
+style guide. Note that archived versions of the style guide may not necessarily
+document formatting for newer Rust constructs that did not exist at the time
+that version of the style guide was archived.
```

Comment:
objection: I'm not confident that this will end up being the best way to track the different style editions and I'd prefer not to specify the specific mechanism as part of this RFC and rather wait until we've tried splitting the style guide for two editions to see what approach we end up with.

```suggestion
The style team will maintain distinct versions of the style guide for each style
edition. Note that archived versions of the style guide may not necessarily
document formatting for newer Rust constructs that did not exist at the time
that version of the style guide was archived.
```